mount_union.py

```diff
@@ -39,8 +39,13 @@ def mount_union(ctx):
         os.mkdir(workdir)
 
         mntopt = " -orw" + cfg.mntopts()
-        system("mount -t overlay overlay " + union_mntroot + mntopt +
-               ",lowerdir=" + lower_mntroot + ",upperdir=" + upperdir + ",workdir=" + workdir)
+        mount_program = cfg.mount_program()
+        if mount_program is None:
+            system("mount -t overlay overlay " + union_mntroot + mntopt +
+                   ",lowerdir=" + lower_mntroot + ",upperdir=" + upperdir + ",workdir=" + workdir)
+        else:
+            system(mount_program + " " + mntopt +
+                   ",lowerdir=" + lower_mntroot + ",upperdir=" + upperdir + ",workdir=" + workdir + " " + union_mntroot)
```

Comment:
wouldn't it look something like mount -t fuse fuse-overlayfs?  Should I assume the type is always `fuse` when a fstype is specified?

Reply:
There is no need for that.
./run --fs=fuse-overlayfs --xdev --verify
is shorter and sufficient

elif args[0] == "--ov":
    cfg.set_testing_overlayfs()
...
elif args[0].startswith("--fs="):
    cfg.set_testing_overlayfs()
    s = args[0]
    fs = s[s.rfind("=")+1:]
    cfg.set_fstype(fs)

...
            system("mount -t " + cfg.fstype() + " overlay " + curr_snapshot + mntopt +
...

that's it.

When mount(8) can't find $fstype in kernel it will try to execute mount.$fstype
with argument <dev> <mnt> <opts>

So I think you only need to change expected arg order in your program to
be compliant as mount helper.

Since your program is doing the mount, there is no need AFAIK for the
secondary fstype option (i.e. fuse.overlayfs-fuse). IIUC this is for fuse
high level implementation that just runs the deamon to register the fuse
filesystem and require user to do the mount, so when user does
mount -t fuse.overlayfs <dev> <mnt> <opt>
the fuse mount helper mount.fuse can bind that to a deamon that
registered as fuse filesystem of type "overlayfs".

But I am not sure about the role of mount.fuse helper, so you need to check that what I am saying really works as I described.

Reply:
... and please update README with example of running with alt fstype

Reply:
having only this option won't work with fuse file systems.  The mount helper for fuse is `mount.fuse` and it uses the spec to find the program to execute:

`mount -t fuse fuse-overlayfs ...`

Having `mount.fuse-overlayfs` doesn't seem like what other FUSE file systems do, ssh for instance installs only `/usr/bin/sshfs` not a `mount.ssh` helper.

Alternatively, I would have to change the program to work differently if argv[0] is `mount.overlayfs` and drop the first argument (`fuse_parse_cmdline` expects only the mount point not the additional spec)

Reply:
I see, but still think only need to specify the compound fstype, i.e.:
./run --fs=fuse.fuse-overlayfs --xdev --verify
Then mount.fuse helper will run
fuse.fuse-overlayfs overlay $mnt $opts

This looks to me like the syntax that other fuse mount helpers use, e.g.:
sshfs [user@]host:[dir] mountpoint [options]

And then, like sshfs, you can also support mounting fuse-overlayfs from /etc/fstab.

I understand that the first argument (fsname) is not needed for fuse-overlayfs
but unionmount uses the arbitrary name "overlay" for fsname and relies on that
name for cleanup of old mounts in set_up() (i.e. grep -q 'overlay ...' /proc/mounts).

From my perspective, the unionmount testsuite support for alternative filesystems
needs to be standard and the syntax above looks like the standard to me.

There is also value for fuse-overlayfs to tag the specific mount with a name so
/proc/mounts has a more accurate description of the setup.